### ISSUE TITLE
Convert 'if' to 'switch' code fix suggested for comparisons with implicit casts

### ISSUE THREAD

**Issue by GGG-KILLER** [dotnet/roslyn#48832]
**Version Used**: 
- .NET 5.0.100-rc.2.20479.15
- VS Version 16.8.0 Preview 4.0

I wasn't able to test in sharplab as I am not even sure if it's supposed to generate these code fixes.

**Steps to Reproduce**:
1. Write a sequence of `if`s where the variable is implicitly converted like the following:
```cs
public static class Duration
{
    public const double TicksPerSecond = TimeSpan.TicksPerHour / 3600D;
    public const double TicksPerMillisecond = TicksPerSecond / 1000D;
    public const double TicksPerMicrosecond = TicksPerMillisecond / 1000D;
    public const double TicksPerNanosecond = TicksPerMicrosecond / 1000D;

    public static string Format(long ticks, string format = "{0:##00.00}{1}")
    {
        if (ticks > TicksPerSecond)
            return string.Format(format, ticks / TicksPerSecond, "s");
        else if (ticks > TicksPerMillisecond)
            return string.Format(format, ticks / TicksPerMillisecond, "ms");
        else if (ticks > TicksPerMicrosecond)
            return string.Format(format, ticks / TicksPerMicrosecond, "μs");
        else
            return string.Format(format, ticks / TicksPerNanosecond, "ns");
    }
}
```
2. Hover over the first `if`.

**Expected Behavior**:
Both "Convert to 'switch' statement" and "Convert to 'switch' expression" code fixes aren't raised or `ticks` is casted to a `double` in the `switch` argument.

**Actual Behavior**:
Following code is generated:
1. `switch` statement code fix:
```cs
switch ( ticks )
{
    case > TicksPerSecond:
        return String.Format ( format, ticks / TicksPerSecond, "s" );
    case > TicksPerMillisecond:
        return String.Format ( format, ticks / TicksPerMillisecond, "ms" );
    case > TicksPerMicrosecond:
        return String.Format ( format, ticks / TicksPerMicrosecond, "μs" );
    default:
        return String.Format ( format, ticks / TicksPerNanosecond, "ns" );
}
```
2. `switch` expression code fix:
```cs
return ticks switch
{
    > TicksPerSecond => String.Format ( format, ticks / TicksPerSecond, "s" ),
    > TicksPerMillisecond => String.Format ( format, ticks / TicksPerMillisecond, "ms" ),
    > TicksPerMicrosecond => String.Format ( format, ticks / TicksPerMicrosecond, "μs" ),
    _ => String.Format ( format, ticks / TicksPerNanosecond, "ns" )
};
```

Both code fix results generate errors because the `double` `const`s aren't implicitly convertible to `long` but the inverse happens in the `if` version.

**Comment by CyrusNajmabadi** [comment]
@alrz i'm interested in fixing this, but i'm not completely sure the best place to do it.  It seems like we need to likely convert the code to:

```
switch ( (double)ticks )
{
```

However, we should only do this (and allow the conversion) if that's how `ticks` is being converted anyways in all the relational checks.  Do you have a suggestion on where to best do that analysis and flow this data through?

THanks!

**Comment by alrz** [comment]
When we check for constants, we could capture the conversion type as the "input type" for AnalyzedPattern (currently we're checking the syntax so it's lost).

After that we have a collection of AnalyzedPatterns in AnalyzedSwitchSection which could be verified if all label patterns have the same type. If not, we bail, otherwise a cast should be generated where necessary.

**Comment by alrz** [comment]
Considering an input type for each pattern could make it possible to do this even if not all patterns require a conversion.
```cs
switch ( ticks )
{
    case > (long)doubleConstant: ..
    case > longConstant: ..
}
```
So if *all* of the arms need a cast, it could be moved to the target expression. That's assuming it doesn't matter which operand is being casted.

**Comment by CyrusNajmabadi** [comment]
> Considering an input type for each pattern could make it possible to do this even if not all patterns require a conversion.

Note: i don't believe those semantics are correct.  You are narrowing the double to a long, whereas the original semantics widen the long to a double.

**Comment by CyrusNajmabadi** [comment]
> When we check for constants, we could capture the conversion type as the "input type" for AnalyzedPattern (currently we're checking the syntax so it's lost).

> After that we have a collection of AnalyzedPatterns in AnalyzedSwitchSection which could be verified if all label patterns have the same type. If not, we bail, otherwise a cast should be generated where necessary.

That sounds reasonable.  I will try to see if i can do that.  Thanks!

**Comment by CyrusNajmabadi** [comment]
I don't repro this:

![image](https://user-images.githubusercontent.com/4564579/199354919-443df80d-0ec6-4372-8103-54ef7ec6d616.png)
